sbol_utilities/sbol3_sbol2_conversion.py

```diff
@@ -175,24 +175,36 @@ def visit_component(self, cp3: sbol3.Component):
         types2 = [type_map.get(t, t) for t in cp3.types]
-        # Make the Component object and add it to the document
-        cp2 = sbol2.ComponentDefinition(self._sbol2_identity(cp3), types2, version=self._sbol2_version(cp3))
-        self.doc2.addComponentDefinition(cp2)
-        # Convert the Component properties not covered by the constructor
-        cp2.roles = cp3.roles
-        cp2.sequences = cp3.sequences
-        if cp3.features:
-            raise NotImplementedError('Conversion of Component features from SBOL3 to SBOL2 not yet implemented')
-        if cp3.interactions:
-            raise NotImplementedError('Conversion of Component interactions from SBOL3 to SBOL2 not yet implemented')
-        if cp3.constraints:
-            raise NotImplementedError('Conversion of Component constraints from SBOL3 to SBOL2 not yet implemented')
-        if cp3.interface:
-            raise NotImplementedError('Conversion of Component interface from SBOL3 to SBOL2 not yet implemented')
-        if cp3.models:
-            raise NotImplementedError('Conversion of Component models from SBOL3 to SBOL2 not yet implemented')
-        # Map over all other TopLevel properties and extensions not covered by the constructor
-        self._convert_toplevel(cp3, cp2)
+
+        # Determine whether Component maps to a ComponentDefinition or ModuleDefinition
+        if sbol3.SBO_FUNCTIONAL_ENTITY not in cp3.types:
+            # Make the Component object and add it to the document
+            cp2 = sbol2.ComponentDefinition(self._sbol2_identity(cp3), types2, version=self._sbol2_version(cp3))
+
+            # Convert the Component properties not covered by the constructor
+            cp2.roles = cp3.roles
+            cp2.sequences = cp3.sequences
+            for f in cp3.features:
+                raise NotImplementedError('Conversion of Component features from SBOL3 to SBOL2 not yet implemented')
+            if cp3.interactions:
+                raise NotImplementedError('Conversion of Component interactions from SBOL3 to SBOL2 not yet implemented')
+            if cp3.constraints:
+                raise NotImplementedError('Conversion of Component constraints from SBOL3 to SBOL2 not yet implemented')
+            if cp3.interface:
+                raise NotImplementedError('Conversion of Component interface from SBOL3 to SBOL2 not yet implemented')
+            if cp3.models:
+                raise NotImplementedError('Conversion of Component models from SBOL3 to SBOL2 not yet implemented')
+            self.doc2.addComponentDefinition(cp2)
+            self._convert_toplevel(cp3, cp2)
+        else:
+            # Component maps to an SBOL2 ModuleDefinition
+            mdef2 = sbol2.ModuleDefinition(self._sbol2_identity(cp3),
+                                           version=self._sbol2_version(cp3))
+            mdef2.roles = cp3.roles
+            mdef2.models = cp3.models
```

Comment:
I don't believe that this is safe: the models need to be converted. Replace this with a `NotImplementedError` like above?

Reply:
That's why I would tend to leave it as "NotImplementedError" for now (like it is above) rather than copy unconverted materials, which will cause errors that are likely to be harder for a user to interpret.

Reply:
Unfortunately, until #326 is resolved, errors from unconverted URIs are inevitable.  Currently we initialize some reference properties by construction, e.g., `ComponentDefinition.sequences` and `Implementation.built`.   Since these properties are initialized by the constructor and they are required, there's no way to raise a `NotImplementedError`.  So, you're stuck with an unconverted URI.   I'd rather just go ahead and implement the copy now rather then leave it for later.